File: Code/05_alzheimers_disease/matlab/ad_continuum.m
clear; clc; close all;
addpath('../../common/matlab');
figDir = fullfile('..','..','..','Report','Figs');
dataDir = fullfile('..','..','..','data_for_section3&4');
groups = {'CN','EMCI','LMCI','AD'};
twilightCols = twilight_colors(4);
cols = {twilightCols(1,:), twilightCols(2,:), twilightCols(3,:), twilightCols(4,:)};
sigmoid = @(x,c,w,lo,hi) lo + (hi-lo) ./ (1 + exp(-(x-c)./w));

data = cell(size(groups));
counts = zeros(size(groups));
for g = 1:numel(groups)
    files = dir(fullfile(dataDir, groups{g}, '*.csv'));
    mats = {};
    minN = inf;
    for i = 1:numel(files)
        try
            M = readmatrix(fullfile(files(i).folder, files(i).name));
            if size(M,1) == size(M,2) && size(M,1) >= 60
                M(isnan(M)) = 0;
                M = log1p(M);
                M = (M + M') / 2;
                mats{end+1} = M; %#ok<SAGROW>
                minN = min(minN, size(M,1));
            end
        catch
        end
    end
    counts(g) = numel(mats);
    stack = zeros(minN, minN, counts(g));
    for i = 1:counts(g)
        stack(:,:,i) = mats{i}(1:minN, 1:minN);
    end
    data{g} = stack;
end

means = cell(size(groups));
for g = 1:numel(groups)
    means{g} = mean(data{g}, 3);
    if max(means{g}(:)) > 0, means{g} = means{g} / max(means{g}(:)); end
    means{g} = means{g} - diag(diag(means{g}));
end

%% Figure 11
fig = figure('Position', [100 100 820 225], 'Color', 'w');
for g = 1:numel(groups)
    subplot(1,4,g);
    imagesc(means{g}, [0 1]); axis image off; colormap(gca, turbo);
    title(sprintf('%s (n=%d)', groups{g}, counts(g)), 'FontSize', 9);
    text(-0.18, 1.05, char('A'+g-1), 'Units', 'normalized', 'FontSize', 11);
end
cb = colorbar('Position', [0.93 0.22 0.015 0.58]);
cb.Label.String = 'Normalized structural connectivity';
apply_kuramoto_style(fig);
exportgraphics(fig, fullfile(figDir, '11_ad_connectome_evolution.png'), 'Resolution', 300, 'BackgroundColor', 'w');

%% Metrics for Figure 12
CN = means{1};
n = size(CN,1);
rawDiffs = [];
lesionRaw = cell(1,3);
for g = 2:4
    lc = sum(CN - means{g}, 2);
    lesionRaw{g-1} = lc;
    rawDiffs = [rawDiffs; lc]; %#ok<AGROW>
end
globalMaxLesion = max(rawDiffs);
lesion = cell(1,3);
for g = 2:4
    lesion{g-1} = lesionRaw{g-1} / globalMaxLesion;
end

[V, ~] = eigs(CN, 1, 'largestabs');
eigCent = abs(V);
Wc = CN.^(1/3);
clust = zeros(n,1);
for i = 1:n
    k_i = sum(CN(i,:) > 0);
    if k_i > 1
        triangles = Wc(i,:) * Wc * Wc(:,i);
        clust(i) = triangles / (k_i * (k_i - 1));
    end
end
threshold = prctile(CN(:), 70);
Gbin = double(CN > threshold);
localEff = zeros(n,1);
for i = 1:n
    neighbors = find(Gbin(i,:));
    k_i = length(neighbors);
    if k_i > 1
        subA = Gbin(neighbors, neighbors);
        D = distances(graph(subA));
        invD = 1 ./ D;
        invD(isinf(invD) | isnan(invD)) = 0;
        invD(1:size(invD,1)+1:end) = 0;
        localEff(i) = sum(invD(:)) / (k_i * (k_i - 1));
    end
end
Dmat = diag(sum(Gbin, 2));
L = Dmat - Gbin;
[Veig, Dval] = eig(L);
eigVals = diag(Dval);
[~, sortIdx] = sort(eigVals);
v2 = Veig(:, sortIdx(2));
v3 = Veig(:, sortIdx(3));
communities = zeros(n,1);
communities(v2 > 0 & v3 > 0) = 1;
communities(v2 > 0 & v3 <= 0) = 2;
communities(v2 <= 0 & v3 > 0) = 3;
communities(v2 <= 0 & v3 <= 0) = 4;
part = zeros(n,1);
for i = 1:n
    k_i = sum(CN(i,:));
    if k_i > 0
        s = 0;
        for c = 1:4
            comNodes = find(communities == c);
            s = s + (sum(CN(i, comNodes)) / k_i)^2;
        end
        part(i) = 1 - s;
    end
end
metrics = {eigCent, clust, localEff, part};
metricNames = {'Eigenvector Centrality','Clustering Coefficient','Local Efficiency','Participation Coefficient'};

fig = figure('Position', [100 100 740 520], 'Color', 'w');
for m = 1:4
    subplot(2,2,m); hold on;
    x = metrics{m};
    leg = cell(1,3);
    for g = 2:4
        y = lesion{g-1};
        scatter(x, y, 18, cols{g}, 'filled', 'MarkerFaceAlpha', 0.45);
        p = polyfit(x, y, 1);
        xx = linspace(min(x), max(x), 80);
        plot(xx, polyval(p, xx), 'Color', cols{g}, 'LineWidth', 1.7);
        R = corrcoef(x, y);
        r2 = R(1,2)^2;
        df2 = length(x) - 2;
        if r2 < 1
            Fstat = (r2 * df2) / (1 - r2);
        else
            Fstat = inf;
        end
        pval = fcdf(Fstat, 1, df2, 'upper');
        if pval < 0.001
            pstr = sprintf('%.2e', pval);
        else
            pstr = sprintf('%.3f', pval);
        end
        leg{g-1} = sprintf('%s: F=%.1f, p=%s', groups{g}, Fstat, pstr);
    end
    text(-0.13, 1.03, char('A'+m-1), 'Units', 'normalized', 'FontSize', 11);
    xlabel(['Baseline ' metricNames{m}]); ylabel('Lesion Count (Normalised)');
    legend(leg, 'Location', 'northwest', 'FontSize', 6.7);
end
apply_kuramoto_style(fig);
exportgraphics(fig, fullfile(figDir, '12_topological_lesion_regression.png'), 'Resolution', 300, 'BackgroundColor', 'w');

%% Figure 13
lambda = linspace(0, 0.05, 151);
R13 = zeros(numel(groups), numel(lambda));
for k = 1:numel(lambda)
    for g = 1:numel(groups)
        R13(g,k) = eval_kuramoto_original(means{g}, lambda(k), 2.5, 0.2);
    end
end
fig = figure('Position', [100 100 545 310], 'Color', 'w');
hold on;
patch([0.015 0.030 0.030 0.015], [0 0 1.05 1.05], [0.16 0.65 0.72], 'FaceAlpha', 0.10, 'EdgeColor', 'none', 'DisplayName', 'Critical Regime (0.01-0.03)');
for g = 1:4
    plot(lambda, R13(g,:), 'Color', cols{g}, 'LineWidth', 1.5, 'DisplayName', sprintf('%s (n=%d)', groups{g}, counts(g)));
end
xlabel('Cortical Coupling Factor, \lambda'); ylabel('Order Parameter r');
xlim([0 0.05]); ylim([0 1.05]);
legend('Location', 'southeast');
grid on; set(gca, 'GridLineStyle', '--', 'GridAlpha', 0.4, 'Box', 'off', 'TickDir', 'out', 'LineWidth', 1.5);
exportgraphics(fig, fullfile(figDir, '13_ad_phase_transition_delay.png'), 'Resolution', 300, 'BackgroundColor', 'w');

%% Figure 14
lambda14 = 0:0.01:0.6;
Mcn84 = load_mat_secure_local(fullfile(dataDir, 'CN'), 84);
Mad84 = load_mat_secure_local(fullfile(dataDir, 'AD'), 84);
[Vhub, ~] = eigs(Mcn84, 1, 'largestabs');
metricEigen = abs(Vhub);
[~, hubIdx] = sort(metricEigen, 'descend');
numHubs = round(0.1 * 84);
topHubs = hubIdx(1:numHubs);
Mhub = Mad84;
Mhub(topHubs,:) = Mcn84(topHubs,:);
Mhub(:,topHubs) = Mcn84(:,topHubs);
rng(42);
randNodes = randperm(84, numHubs);
Mrand = Mad84;
Mrand(randNodes,:) = Mcn84(randNodes,:);
Mrand(:,randNodes) = Mcn84(:,randNodes);
rng(100);
omegaInit = randn(84,1) * 0.25;
thetaInit = rand(84,1) * 2*pi;
Rcn = zeros(size(lambda14)); Rad = Rcn; Rhub = Rcn; Rrand = Rcn;
for k = 1:numel(lambda14)
    Rcn(k) = kuramoto_rescue_original(Mcn84, lambda14(k), omegaInit, thetaInit);
    Rad(k) = kuramoto_rescue_original(Mad84, lambda14(k), omegaInit, thetaInit);
    Rhub(k) = kuramoto_rescue_original(Mhub, lambda14(k), omegaInit, thetaInit);
    Rrand(k) = kuramoto_rescue_original(Mrand, lambda14(k), omegaInit, thetaInit);
end
fig = figure('Position', [100 100 545 325], 'Color', 'w');
hold on;
plot(lambda14, Rcn, '-^', 'Color', cols{1}, 'LineWidth', 1.5, 'MarkerSize', 3.2, 'MarkerFaceColor', cols{1}, 'DisplayName', 'Healthy Baseline (CN)');
plot(lambda14, Rhub, '-o', 'Color', cols{3}, 'LineWidth', 1.5, 'MarkerSize', 3.6, 'MarkerFaceColor', cols{3}, 'DisplayName', 'Targeted Rescue (Top 10 Hubs)');
plot(lambda14, Rrand, '--s', 'Color', cols{2}, 'LineWidth', 1.5, 'MarkerSize', 3.2, 'MarkerFaceColor', cols{2}, 'DisplayName', 'Random Rescue (10 Non-Hubs)');
plot(lambda14, Rad, ':', 'Color', cols{4}, 'LineWidth', 1.5, 'DisplayName', 'Pathological Baseline (AD)');
xlabel('Coupling Strength, \lambda'); ylabel('Global Synchronization, R');
xlim([0 0.6]); ylim([0 1.05]);
legend('Location', 'southeast', 'Box', 'off');
box off; grid on; set(gca, 'LineWidth', 1.5, 'TickDir', 'out', 'GridAlpha', 0.15);
exportgraphics(fig, fullfile(figDir, '14_perturbation_rescue_experiment.png'), 'Resolution', 300, 'BackgroundColor', 'w');

function rGlobal = eval_kuramoto_original(M, lam, omegaStd, noiseStrength)
rng(42);
N = size(M, 1);
omega = randn(N, 1) * omegaStd;
theta = rand(N, 1) * 2*pi;
tMax = 40;
dt = 0.01;
steps = floor(tMax / dt);
steadySteps = floor(steps / 2);
thetaSteady = zeros(N, steadySteps);
noiseFactor = noiseStrength * sqrt(dt);
for step = 1:steps
    phaseDiff = theta' - theta;
    dtheta = omega + 80 * lam * sum(M .* sin(phaseDiff), 2);
    theta = theta + dtheta * dt + noiseFactor * randn(N, 1);
    if step > steadySteps
        thetaSteady(:, step - steadySteps) = theta;
    end
end
rGlobal = mean(abs(mean(exp(1i * thetaSteady), 1)));
end

function R = kuramoto_rescue_original(M, lambdaVal, omegaInit, thetaInit)
rng(42);
noiseStrength = 0.25;
dt = 0.01;
tMax = 30;
steps = floor(tMax / dt);
steadyStart = floor(steps * 0.6);
omega = omegaInit;
theta = thetaInit;
noiseFactor = noiseStrength * sqrt(dt);
Rvec = zeros(1, steps - steadyStart + 1);
idx = 1;
for step = 1:steps
    phaseDiff = theta' - theta;
    drift = omega + lambdaVal * sum(M .* sin(phaseDiff), 2);
    theta = theta + drift * dt + noiseFactor * randn(size(theta));
    if step >= steadyStart
        Rvec(idx) = abs(mean(exp(1i * theta)));
        idx = idx + 1;
    end
end
R = mean(Rvec);
end

function meanMat = load_mat_secure_local(folder, N)
files = dir(fullfile(folder, '*.csv'));
sumMat = zeros(N, N);
validCount = 0;
for j = 1:length(files)
    try
        temp = readmatrix(fullfile(folder, files(j).name));
        if isempty(temp) || size(temp, 1) < N || size(temp, 2) < N, continue; end
        mat = temp(end-N+1:end, end-N+1:end);
        mat = log1p(mat);
        mat = (mat + mat') / 2;
        sumMat = sumMat + mat;
        validCount = validCount + 1;
    catch
        continue;
    end
end
meanMat = sumMat / validCount;
if max(meanMat(:)) > 0, meanMat = meanMat / max(meanMat(:)); end
meanMat(1:N+1:end) = 0;
end

function cmap = magma_or_hot()
try
    cmap = magma(256);
catch
    cmap = hot(256);
end
end

function y = clip01(y)
y = max(0, min(1, y));
end

function cols = twilight_colors(n)
base = [ ...
    0.8858 0.8500 0.8879
    0.6443 0.4390 0.7389
    0.1849 0.0794 0.2131
    0.2453 0.2878 0.5373
    0.2890 0.5650 0.7595
    0.7590 0.7060 0.5680
    0.8858 0.8500 0.8879];
x = linspace(0, 1, size(base, 1));
xi = linspace(0.08, 0.88, n);
cols = interp1(x, base, xi);
end

File: Code/05_alzheimers_disease/python/ad_continuum.py
from __future__ import annotations

import glob
import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from scipy import stats
from sklearn.linear_model import LinearRegression

sys.path.append(str(Path(__file__).resolve().parents[2] / "common" / "python"))
from kuramoto_style import COLORS, DATA_DIR, clean_axes, panel_label, save_figure, smooth_transition


GROUPS = ["CN", "EMCI", "LMCI", "AD"]
GROUP_COLORS = dict(zip(GROUPS, plt.cm.twilight(np.linspace(0.08, 0.88, len(GROUPS)))))


def load_csv_matrix(path: str) -> np.ndarray | None:
    try:
        mat = pd.read_csv(path, header=None).to_numpy(float)
    except Exception:
        return None
    if mat.ndim != 2 or mat.shape[0] != mat.shape[1] or mat.shape[0] < 60:
        return None
    mat = np.nan_to_num(mat)
    mat = 0.5 * (mat + mat.T)
    np.fill_diagonal(mat, 0)
    if mat.max() > 0:
        mat = mat / mat.max()
    return mat


def load_group_matrices():
    data = {}
    for group in GROUPS:
        mats = []
        for path in glob.glob(str(DATA_DIR / group / "*.csv")):
            mat = load_csv_matrix(path)
            if mat is not None:
                mats.append(mat)
        n = min(m.shape[0] for m in mats)
        mats = np.asarray([m[:n, :n] for m in mats])
        data[group] = mats
    return data


def load_original_group_means(*, n_nodes: int | None = None, tail: bool = False):
    means = {}
    counts = {}
    for group in GROUPS:
        paths = sorted(glob.glob(str(DATA_DIR / group / "*.csv")))
        sum_mat = None
        valid = 0
        for path in paths:
            try:
                mat = pd.read_csv(path, header=None).to_numpy(float)
            except Exception:
                continue
            if mat.ndim != 2 or mat.shape[0] != mat.shape[1]:
                continue
            if n_nodes is not None:
                if mat.shape[0] < n_nodes:
                    continue
                mat = mat[-n_nodes:, -n_nodes:] if tail else mat[:n_nodes, :n_nodes]
            mat = np.log1p(np.nan_to_num(mat))
            mat = 0.5 * (mat + mat.T)
            if sum_mat is None:
                sum_mat = np.zeros_like(mat, dtype=float)
            if mat.shape != sum_mat.shape:
                continue
            sum_mat += mat
            valid += 1
        if valid:
            mean_mat = sum_mat / valid
            if mean_mat.max() > 0:
                mean_mat = mean_mat / mean_mat.max()
            np.fill_diagonal(mean_mat, 0)
            means[group] = mean_mat
            counts[group] = valid
    return means, counts


def binary(mean_mat, density=0.18):
    tri = mean_mat[np.triu_indices_from(mean_mat, 1)]
    cutoff = np.quantile(tri[tri > 0], 1 - density)
    a = (mean_mat >= cutoff).astype(float)
    np.fill_diagonal(a, 0)
    return a


def graph_metrics(a):
    degree = a.sum(axis=1)
    w, v = np.linalg.eigh(a)
    eig = np.abs(v[:, np.argmax(w)])
    eig = eig / eig.max()

    n = a.shape[0]
    clustering = np.zeros(n)
    local_eff = np.zeros(n)
    for i in range(n):
        nb = np.where(a[i] > 0)[0]
        k = len(nb)
        if k >= 2:
            sub = a[np.ix_(nb, nb)]
            clustering[i] = sub.sum() / (k * (k - 1))
            local_eff[i] = clustering[i] / np.sqrt(k)
    modules = np.array_split(np.arange(n), 5)
    participation = np.zeros(n)
    for i in range(n):
        ki = max(degree[i], 1)
        participation[i] = 1 - sum((a[i, mod].sum() / ki) ** 2 for mod in modules)
    return {
        "Eigenvector Centrality": eig,
        "Clustering Coefficient": clustering,
        "Local Efficiency": local_eff,
        "Participation Coefficient": participation,
        "Degree": degree,
    }


def connectome_evolution(data):
    means = {g: data[g].mean(axis=0) for g in GROUPS}
    fig, axes = plt.subplots(1, 4, figsize=(8.2, 2.25))
    for i, g in enumerate(GROUPS):
        ax = axes[i]
        im = ax.imshow(means[g], cmap="turbo", vmin=0, vmax=1, interpolation="nearest")
        panel_label(ax, chr(ord("A") + i), -0.18, 1.05)
        ax.set_title(f"{g} (n={len(data[g])})", fontsize=9)
        ax.set_xticks([])
        ax.set_yticks([])
    cbar = fig.colorbar(im, ax=axes, fraction=0.030, pad=0.015)
    cbar.set_label("Normalized structural connectivity")
    save_figure(fig, "11_ad_connectome_evolution.png")


def lesion_counts(means):
    cn = means["CN"]
    counts = {}
    for g in ["EMCI", "LMCI", "AD"]:
        loss = np.maximum(cn - means[g], 0)
        counts[g] = loss.sum(axis=1)
        if counts[g].max() > 0:
            counts[g] = counts[g] / counts[g].max()
    return counts


def lesion_regression(data):
    means, _ = load_original_group_means()
    cn = means["CN"]
    disease_groups = ["EMCI", "LMCI", "AD"]
    raw_diffs = {g: (cn - means[g]).sum(axis=1) for g in disease_groups}
    global_max = max(float(v.max()) for v in raw_diffs.values())
    lesions = {g: raw_diffs[g] / global_max for g in disease_groups}

    vals, vecs = np.linalg.eigh(cn)
    eigen = np.abs(vecs[:, np.argmax(vals)])

    w_cbrt = np.cbrt(cn)
    n = cn.shape[0]
    cluster = np.zeros(n)
    for i in range(n):
        k_i = np.sum(cn[i] > 0)
        if k_i > 1:
            triangles = w_cbrt[i, :] @ w_cbrt @ w_cbrt[:, i]
            cluster[i] = triangles / (k_i * (k_i - 1))

    threshold = np.percentile(cn, 70)
    g_bin = (cn > threshold).astype(float)
    efficiency = np.zeros(n)
    for i in range(n):
        neighbors = np.where(g_bin[i] > 0)[0]
        k_i = len(neighbors)
        if k_i > 1:
            sub = g_bin[np.ix_(neighbors, neighbors)]
            dist = shortest_paths_unweighted(sub)
            inv = np.divide(1.0, dist, out=np.zeros_like(dist), where=np.isfinite(dist) & (dist > 0))
            efficiency[i] = inv.sum() / (k_i * (k_i - 1))

    degree = np.diag(g_bin.sum(axis=1))
    lap = degree - g_bin
    eig_vals, eig_vecs = np.linalg.eigh(lap)
    order = np.argsort(eig_vals)
    v2 = eig_vecs[:, order[1]]
    v3 = eig_vecs[:, order[2]]
    communities = np.zeros(n, dtype=int)
    communities[(v2 > 0) & (v3 > 0)] = 1
    communities[(v2 > 0) & (v3 <= 0)] = 2
    communities[(v2 <= 0) & (v3 > 0)] = 3
    communities[(v2 <= 0) & (v3 <= 0)] = 4
    participation = np.zeros(n)
    for i in range(n):
        k_i = cn[i].sum()
        if k_i > 0:
            participation[i] = 1 - sum((cn[i, communities == c].sum() / k_i) ** 2 for c in range(1, 5))

    metrics = {
        "Eigenvector Centrality": eigen,
        "Clustering Coefficient": cluster,
        "Local Efficiency": efficiency,
        "Participation Coefficient": participation,
    }
    metric_names = list(metrics)
    fig, axes = plt.subplots(2, 2, figsize=(7.4, 5.2))
    axes = axes.ravel()
    for idx, name in enumerate(metric_names):
        ax = axes[idx]
        x = metrics[name]
        handles = []
        labels = []
        for g in disease_groups:
            y = lesions[g]
            ax.scatter(x, y, s=18, color=GROUP_COLORS[g], alpha=0.45, edgecolors="none")
            slope, intercept = np.polyfit(x, y, 1)
            r = np.corrcoef(x, y)[0, 1]
            r_squared = r**2
            df_2 = len(x) - 2
            f_stat = (r_squared * df_2) / (1 - r_squared) if r_squared < 1 else np.inf
            p = stats.f.sf(f_stat, 1, df_2)
            xx = np.linspace(x.min(), x.max(), 100)
            h, = ax.plot(xx, slope * xx + intercept, lw=1.7, color=GROUP_COLORS[g])
            handles.append(h)
            p_str = f"{p:.2e}" if p < 0.001 else f"{p:.3f}"
            labels.append(f"{g}: F={f_stat:.1f}, p={p_str}")
        panel_label(ax, chr(ord("A") + idx), -0.13, 1.03)
        clean_axes(ax, f"Baseline {name}", "Lesion Count (Normalised)")
        ax.legend(handles, labels, frameon=False, loc="upper left", fontsize=6.7)
    fig.subplots_adjust(wspace=0.32, hspace=0.38)
    save_figure(fig, "12_topological_lesion_regression.png")


def shortest_paths_unweighted(adj):
    n = adj.shape[0]
    dist = np.full((n, n), np.inf)
    for source in range(n):
        dist[source, source] = 0
        frontier = [source]
        while frontier:
            current = frontier.pop(0)
            for nb in np.where(adj[current] > 0)[0]:
                if not np.isfinite(dist[source, nb]):
                    dist[source, nb] = dist[source, current] + 1
                    frontier.append(nb)
    return dist


def kuramoto_single(mean_mat, lam, omega, theta_init, *, noise_strength, scale, steps, burn, dt, noise_seed=42):
    rng = np.random.default_rng(noise_seed)
    n = mean_mat.shape[0]
    mat = np.asarray(mean_mat, dtype=float)
    np.fill_diagonal(mat, 0)
    theta = theta_init.copy()
    r_trace = []
    noise_factor = noise_strength * np.sqrt(dt)
    for step in range(steps):
        phase = theta[None, :] - theta[:, None]
        coupling = (mat * np.sin(phase)).sum(axis=1)
        drift = omega + scale * lam * coupling
        theta += drift * dt + noise_factor * rng.normal(size=n)
        if step >= burn:
            r_trace.append(abs(np.exp(1j * theta).mean()))
    return float(np.mean(r_trace))


def transition_curve_original(mean_mat, lambdas):
    vals = []
    for lam in lambdas:
        rng = np.random.default_rng(42)
        omega = rng.normal(0, 2.5, mean_mat.shape[0])
        theta = rng.uniform(0, 2 * np.pi, mean_mat.shape[0])
        vals.append(kuramoto_single(mean_mat, lam, omega, theta, noise_strength=0.2, scale=80.0, steps=4000, burn=2000, dt=0.01, noise_seed=42))
    return np.asarray(vals)


def rescue_curve_original(mean_mat, lambdas, omega_init, theta_init):
    return np.asarray([
        kuramoto_single(mean_mat, lam, omega_init, theta_init, noise_strength=0.25, scale=1.0, steps=3000, burn=1800, dt=0.01, noise_seed=42)
        for lam in lambdas
    ])


def transition_delay(data):
    lambdas = np.linspace(0, 0.05, 151)
    means, counts = load_original_group_means()
    curves = {g: transition_curve_original(means[g], lambdas) for g in GROUPS}
    fig, ax = plt.subplots(figsize=(5.45, 3.25))
    ax.axvspan(0.015, 0.030, color=COLORS["blue"], alpha=0.10, lw=0, label="Critical Regime (0.01-0.03)")
    for g in GROUPS:
        ax.plot(lambdas, curves[g], lw=1.5, color=GROUP_COLORS[g], label=f"{g} (n={counts[g]})")
    ax.set_xlabel("Cortical Coupling Factor, $\\lambda$")
    ax.set_ylabel("Order Parameter $r$")
    ax.set_xlim(0, 0.05)
    ax.set_ylim(0, 1.05)
    ax.grid(True, linestyle="--", alpha=0.40)
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)
    ax.tick_params(direction="out")
    ax.legend(loc="lower right", frameon=False, fontsize=7.2)
    save_figure(fig, "13_ad_phase_transition_delay.png")


def rescue_experiment(data):
    lambdas = np.arange(0.0, 0.6001, 0.01)
    twilight = plt.cm.twilight(np.linspace(0.08, 0.88, 4))
    means, _ = load_original_group_means(n_nodes=84, tail=True)
    n = 84
    m_cn = means["CN"]
    m_ad = means["AD"]
    vals, vecs = np.linalg.eigh(m_cn)
    eig = np.abs(vecs[:, np.argmax(vals)])
    hubs = np.argsort(eig)[::-1][: round(0.10 * n)]
    rng = np.random.default_rng(42)
    random_nodes = rng.permutation(n)[: len(hubs)]
    m_hub = m_ad.copy()
    m_hub[hubs, :] = m_cn[hubs, :]
    m_hub[:, hubs] = m_cn[:, hubs]
    m_rand = m_ad.copy()
    m_rand[random_nodes, :] = m_cn[random_nodes, :]
    m_rand[:, random_nodes] = m_cn[:, random_nodes]
    init_rng = np.random.default_rng(100)
    omega_init = init_rng.normal(0, 0.25, n)
    theta_init = init_rng.uniform(0, 2 * np.pi, n)
    curves = {
        "Healthy Baseline (CN)": (rescue_curve_original(m_cn, lambdas, omega_init, theta_init), twilight[0], "-^", 1.5, 3.2),
        "Targeted Rescue (Top 10 Hubs)": (rescue_curve_original(m_hub, lambdas, omega_init, theta_init), twilight[2], "-o", 1.5, 3.6),
        "Random Rescue (10 Non-Hubs)": (rescue_curve_original(m_rand, lambdas, omega_init, theta_init), twilight[1], "--s", 1.5, 3.2),
        "Pathological Baseline (AD)": (rescue_curve_original(m_ad, lambdas, omega_init, theta_init), twilight[3], ":", 1.5, 0),
    }
    fig, ax = plt.subplots(figsize=(5.45, 3.5))
    for label, (y, color, style, lw, ms) in curves.items():
        kwargs = {"color": color, "lw": lw, "label": label}
        if ms:
            kwargs.update({"ms": ms, "markerfacecolor": color})
        ax.plot(lambdas, y, style, **kwargs)
    ax.set_xlabel("Coupling Strength, $\\lambda$")
    ax.set_ylabel("Global Synchronization, $R$")
    ax.set_xlim(0, 0.60)
    ax.set_ylim(0, 1.05)
    ax.grid(True, alpha=0.15)
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)
    ax.tick_params(direction="out")
    ax.legend(loc="lower right", frameon=False, fontsize=7)
    save_figure(fig, "14_perturbation_rescue_experiment.png")


if __name__ == "__main__":
    group_data = load_group_matrices()
    connectome_evolution(group_data)
    lesion_regression(group_data)
    transition_delay(group_data)
    rescue_experiment(group_data)
